amgcl/preconditioner/cpr.hpp

```diff
@@ -351,6 +362,91 @@ class cpr {
             rs = backend_type::create_vector(n, bprm);
         }
 
+        void update_transfer(std::shared_ptr<build_matrix> K, const backend_params bprm, std::true_type)
+        {
+            typedef typename backend::row_iterator<build_matrix>::type row_iterator;
+            const int       B = prm.block_size;
+            const ptrdiff_t N = (prm.active_rows ? prm.active_rows : n);
+
+            np = N / B;
+
+            auto fpp = std::make_shared<build_matrix>();
+            fpp->set_size(np, N);
+            fpp->set_nonzeros(N);
+            fpp->ptr[0] = 0;
+
+            // Get the pressure matrix nonzero pattern,
+            // extract and invert block diagonals.
+#pragma omp parallel
+            {
+                std::vector<row_iterator> k; k.reserve(B);
+                multi_array<value_type, 2> v(B, B);
+
+#pragma omp for
+                for(ptrdiff_t ip = 0; ip < static_cast<ptrdiff_t>(np); ++ip) {
+                    ptrdiff_t ik = ip * B;
+                    bool      done = true;
+                    ptrdiff_t cur_col = 0;
+
+                    k.clear();
+                    for(int i = 0; i < B; ++i) {
+                        k.push_back(backend::row_begin(*K, ik + i));
+
+                        if (k.back() && k.back().col() < N) {
+                            ptrdiff_t col = k.back().col() / B;
+                            if (done) {
+                                cur_col = col;
+                                done = false;
+                            } else {
+                                cur_col = std::min(cur_col, col);
+                            }
+                        }
+                        fpp->col[ik + i] = ik + i;
+                    }
+                    fpp->ptr[ip+1] = ik + B;
+
+                    while (!done) {
+                        ptrdiff_t end = (cur_col + 1) * B;
+
+                        if (cur_col == ip) {
+                            // This is diagonal block.
+                            // Capture its (transposed) value,
+                            // invert it and put the relevant row into fpp.
+                            for(int i = 0; i < B; ++i)
+                                for(int j = 0; j < B; ++j) v(i,j) = 0;
+
+                            for(int i = 0; i < B; ++i)
+                                for(; k[i] && k[i].col() < end; ++k[i])
+                                    v(k[i].col() % B, i) = k[i].value();
+
+                            invert(v.data(), &fpp->val[ik]);
```

Comment:
We can put a `break` after this line since we only need the diagonal block to fill out fpp.